### ISSUE TITLE
Error: stat /var/snap/microk8s/common/default-storage/***: no such file or directory

### ISSUE THREAD

**Issue by mikebell** [ubuntu/microk8s#1380]
Hey folks

Setup:

1 master, 3 agents running on PVE.

I'm trying to install the Unifi controller using the following:

`helm install -n unifi unifi stable/unifi`

The pod fails to come up with the following error `CreateContainerConfigError`

```
-> % k -n unifi describe pod unifi-57768dd99b-k67bl
Name:         unifi-57768dd99b-k67bl
Namespace:    unifi
Priority:     0
Node:         192.168.1.15/192.168.1.15
Start Time:   Thu, 02 Jul 2020 14:43:45 +0100
Labels:       app.kubernetes.io/instance=unifi
              app.kubernetes.io/name=unifi
              pod-template-hash=57768dd99b
Annotations:  <none>
Status:       Pending
IP:           10.1.98.48
IPs:
  IP:           10.1.98.48
Controlled By:  ReplicaSet/unifi-57768dd99b
Containers:
  unifi:
    Container ID:
    Image:          jacobalberty/unifi:5.12.35
    Image ID:
    Ports:          8443/TCP, 8080/TCP, 10001/UDP, 3478/UDP
    Host Ports:     0/TCP, 0/TCP, 0/UDP, 0/UDP
    State:          Waiting
      Reason:       CreateContainerConfigError
    Ready:          False
    Restart Count:  0
    Liveness:       http-get https://:https-gui/status delay=30s timeout=1s period=10s #success=1 #failure=3
    Readiness:      http-get https://:https-gui/status delay=15s timeout=1s period=10s #success=1 #failure=3
    Environment:
      UNIFI_HTTP_PORT:   8080
      UNIFI_HTTPS_PORT:  8443
      TZ:                UTC
      RUNAS_UID0:        false
      UNIFI_UID:         999
      UNIFI_GID:         999
    Mounts:
      /unifi/cert from unifi-data (rw,path="cert")
      /unifi/data from unifi-data (rw,path="data")
      /unifi/init.d from unifi-data (rw,path="init.d")
      /unifi/log from unifi-data (rw,path="log")
      /var/run/secrets/kubernetes.io/serviceaccount from default-token-wktzn (ro)
Conditions:
  Type              Status
  Initialized       True
  Ready             False
  ContainersReady   False
  PodScheduled      True
Volumes:
  unifi-data:
    Type:       PersistentVolumeClaim (a reference to a PersistentVolumeClaim in the same namespace)
    ClaimName:  unifi
    ReadOnly:   false
  default-token-wktzn:
    Type:        Secret (a volume populated by a Secret)
    SecretName:  default-token-wktzn
    Optional:    false
QoS Class:       BestEffort
Node-Selectors:  <none>
Tolerations:     node.kubernetes.io/not-ready:NoExecute for 300s
                 node.kubernetes.io/unreachable:NoExecute for 300s
Events:
  Type     Reason     Age                  From                   Message
  ----     ------     ----                 ----                   -------
  Normal   Scheduled  <unknown>            default-scheduler      Successfully assigned unifi/unifi-57768dd99b-k67bl to 192.168.1.15
  Normal   Pulled     43m (x13 over 46m)   kubelet, 192.168.1.15  Container image "jacobalberty/unifi:5.12.35" already present on machine
  Warning  Failed     49s (x210 over 46m)  kubelet, 192.168.1.15  Error: stat /var/snap/microk8s/common/default-storage/unifi-unifi-pvc-3746b7bb-1b91-46a1-9299-f4b6e1c988a1: no such file or directory
```

```
-> % k -n unifi describe persistentvolumeclaims unifi
Name:          unifi
Namespace:     unifi
StorageClass:  microk8s-hostpath
Status:        Bound
Volume:        pvc-3746b7bb-1b91-46a1-9299-f4b6e1c988a1
Labels:        app.kubernetes.io/instance=unifi
               app.kubernetes.io/managed-by=Helm
               app.kubernetes.io/name=unifi
               helm.sh/chart=unifi-0.10.0
Annotations:   control-plane.alpha.kubernetes.io/leader:
                 {"holderIdentity":"72166ba7-b5fc-11ea-9779-eee2ce5a5fd7","leaseDurationSeconds":15,"acquireTime":"2020-07-02T13:43:42Z","renewTime":"2020-...
               meta.helm.sh/release-name: unifi
               meta.helm.sh/release-namespace: unifi
               pv.kubernetes.io/bind-completed: yes
               pv.kubernetes.io/bound-by-controller: yes
               volume.beta.kubernetes.io/storage-provisioner: microk8s.io/hostpath
Finalizers:    [kubernetes.io/pvc-protection]
Capacity:      5Gi
Access Modes:  RWO
VolumeMode:    Filesystem
Mounted By:    unifi-57768dd99b-k67bl
Events:
  Type    Reason                 Age                From                                                       Message
  ----    ------                 ----               ----                                                       -------
  Normal  ExternalProvisioning   46m (x2 over 46m)  persistentvolume-controller                                waiting for a volume to be created, either by external provisioner "microk8s.io/hostpath" or manually created by system administrator
  Normal  Provisioning           46m                microk8s.io/hostpath 72166ba7-b5fc-11ea-9779-eee2ce5a5fd7  External provisioner is provisioning volume for claim "unifi/unifi"
  Normal  ProvisioningSucceeded  46m                microk8s.io/hostpath 72166ba7-b5fc-11ea-9779-eee2ce5a5fd7  Successfully provisioned volume pvc-3746b7bb-1b91-46a1-9299-f4b6e1c988a1
```

```
-> % k describe -n unifi persistentvolume pvc-3746b7bb-1b91-46a1-9299-f4b6e1c988a1
Name:            pvc-3746b7bb-1b91-46a1-9299-f4b6e1c988a1
Labels:          <none>
Annotations:     hostPathProvisionerIdentity: 192.168.1.28
                 pv.kubernetes.io/provisioned-by: microk8s.io/hostpath
Finalizers:      [kubernetes.io/pv-protection]
StorageClass:    microk8s-hostpath
Status:          Bound
Claim:           unifi/unifi
Reclaim Policy:  Delete
Access Modes:    RWO
VolumeMode:      Filesystem
Capacity:        5Gi
Node Affinity:   <none>
Message:
Source:
    Type:          HostPath (bare host directory volume)
    Path:          /var/snap/microk8s/common/default-storage/unifi-unifi-pvc-3746b7bb-1b91-46a1-9299-f4b6e1c988a1
    HostPathType:
Events:            <none>
```

All 4 servers have ~20gb of space left on them. The file exists on Agent 1 see:

![image](https://user-images.githubusercontent.com/242104/86371974-7cdacc00-bc79-11ea-9b3c-e2c3ab0c6929.png)

However it doesn't exist on any of the other servers.

I'm happy to provide more debug output but I'm stumped where to go from here, any help would be greatly appreciated.

[inspection-report-20200702_141305.tar.gz](https://github.com/ubuntu/microk8s/files/4864906/inspection-report-20200702_141305.tar.gz)

Update:

If I understand things correctly then the pod is currently on 192.168.1.15 but the persistent claim is on Agent 1 with is 192.168.1.28 so that's going to cause an issue? Does that assumption sound correct? If so I'm not sure how to fix this.

**Comment by balchua** [comment]
@mikebell yes your analysis is correct.  The default storage in microk8s is a hostpath parocisioner.  Microk8s currently do not have a storage that supports multi nodes.
I tried sometime ago using rook.  But i find it difficult to setup as it needs the host to have free block devices that it can use. Your mileage may vary.

**Comment by anatolbran** [comment]
I get the same issue on a single node cluster
```
  Warning  Failed     75s (x9 over 7m40s)   kubelet            Error: stat /var/snap/microk8s/common/default-storage/default-pod-pvc-pvc-ceddea3b-639a-49d4-94a9-c5a40f563694 no such file or directory
```
Both the pvc and pv show as bound
```
pod-pvc                 Bound    pvc-ceddea3b-639a-49d4-94a9-c5a40f563694   10Gi       RWO            microk8s-persistent-hostpath   89m
```
```
pvc-ceddea3b-639a-49d4-94a9-c5a40f563694   10Gi       RWO            Delete           Bound    default/pod-pvc                 microk8s-persistent-hostpath            89m

```

Anyway to debug this?

**Comment by balchua** [comment]
Is this after reboot or microk8s stop/start?

**Comment by anatolbran** [comment]
> Is this after reboot or microk8s stop/start?

It is actually right at provisioning, I am just creating this cluster
I have 21 apps deployed on it. And all the other 20 work, but this one doesn't
Previously this happened to  other 2  pods, and I deleted and recreated everything just to make sure.  But i see the error persists.
The error-ing pods are different

**Comment by balchua** [comment]
Does the hostpath provisioner logs show something?

**Comment by anatolbran** [comment]
> Does the hostpath provisioner logs show something?

```
I1205 04:00:54.627727       1 leaderelection.go:176] successfully acquired lease to provision for pvc default/pod-pvc
I1205 04:00:54.627900       1 controller.go:893] scheduleOperation[provision-default/pod-pvc[ceddea3b-639a-49d4-94a9-c5a40f563694]]
I1205 04:01:01.602240       1 controller.go:893] scheduleOperation[provision-default/pod-pvc[ceddea3b-639a-49d4-94a9-c5a40f563694]]
I1205 04:01:16.602531       1 controller.go:893] scheduleOperation[provision-default/pod-pvc[ceddea3b-639a-49d4-94a9-c5a40f563694]]
I1205 04:01:27.732901       1 leaderelection.go:202] stopped trying to renew lease to provision for pvc default/pod-pvc, timeout reached
I1205 04:01:31.602746       1 controller.go:893] scheduleOperation[provision-default/pod-pvc[ceddea3b-639a-49d4-94a9-c5a40f563694]]
I1205 04:01:46.603186       1 controller.go:893] scheduleOperation[lock-provision-default/pod-pvc[ceddea3b-639a-49d4-94a9-c5a40f563694]]
I1205 04:01:50.124690       1 leaderelection.go:176] successfully acquired lease to provision for pvc default/pod-pvc
I1205 04:01:50.124871       1 controller.go:893] scheduleOperation[provision-default/pod-pvc[ceddea3b-639a-49d4-94a9-c5a40f563694]]
I1205 04:02:01.603636       1 controller.go:893] scheduleOperation[provision-default/pod-pvc[ceddea3b-639a-49d4-94a9-c5a40f563694]]
I1205 04:02:25.730133       1 leaderelection.go:202] stopped trying to renew lease to provision for pvc default/pod-pvc, timeout reached

```
Sections of these 2 messages repeatedly

**Comment by anatolbran** [comment]
I fixed it by deleting and recreating the problematic pvc, all worked well after that (same code, same process)
I created a snapshot of the server, in case there is need to investigate further please let me know how and what is required and I will gladly provide the necessary debug info

**Comment by joshbranham** [comment]
I am hitting this pretty frequently as well now 

```
E1221 17:49:44.977084       1 leaderelection.go:271] Failed to update lock: Operation cannot be fulfilled on persistentvolumeclaims "kafka-kafka-0": the object has been modified; please apply your changes to the latest version and try again
E1221 17:49:45.377060       1 leaderelection.go:271] Failed to update lock: Operation cannot be fulfilled on persistentvolumeclaims "postgres-postgres-0": the object has been modified; please apply your changes to the latest version and try again
I1221 17:50:13.958360       1 leaderelection.go:202] stopped trying to renew lease to provision for pvc postgres/postgres-postgres-0, timeout reached
I1221 17:50:13.963575       1 leaderelection.go:202] stopped trying to renew lease to provision for pvc elasticsearch/es-data-esnode-0, timeout reached
I1221 17:50:13.967578       1 leaderelection.go:202] stopped trying to renew lease to provision for pvc fakes3/fakes3-fakes3-0, timeout reached
I1221 17:50:13.972383       1 leaderelection.go:202] stopped trying to renew lease to provision for pvc kafka/kafka-kafka-0, timeout reached
I1221 17:50:46.701798       1 controller.go:893] scheduleOperation[lock-provision-monitoring/elasticsearch-data-monitoring-es-default-0[4869d0fe-a082-4e33-bb18-191ce730720d]]
I1221 17:50:46.715064       1 controller.go:695] cannot start watcher for PVC monitoring/elasticsearch-data-monitoring-es-default-0: unknown (get events)
E1221 17:50:46.715096       1 controller.go:534] Error watching for provisioning success, can't provision for claim "monitoring/elasticsearch-data-monitoring-es-default-0": unknown (get events)
I1221 17:50:46.715108       1 leaderelection.go:154] attempting to acquire leader lease...
I1221 17:50:46.720630       1 controller.go:893] scheduleOperation[lock-provision-monitoring/elasticsearch-data-monitoring-es-default-0[4869d0fe-a082-4e33-bb18-191ce730720d]]
E1221 17:50:46.724413       1 leaderelection.go:271] Failed to update lock: Operation cannot be fulfilled on persistentvolumeclaims "elasticsearch-data-monitoring-es-default-0": the object has been modified; please apply your changes to the latest version and try again
I1221 17:50:49.230683       1 leaderelection.go:176] successfully acquired lease to provision for pvc monitoring/elasticsearch-data-monitoring-es-default-0
I1221 17:50:49.230774       1 controller.go:893] scheduleOperation[provision-monitoring/elasticsearch-data-monitoring-es-default-0[4869d0fe-a082-4e33-bb18-191ce730720d]]
I1221 17:50:49.238788       1 hostpath-provisioner.go:86] creating backing directory: /var/snap/microk8s/common/default-storage/monitoring-elasticsearch-data-monitoring-es-default-0-pvc-4869d0fe-a082-4e33-bb18-191ce730720d
I1221 17:50:49.238964       1 controller.go:627] volume "pvc-4869d0fe-a082-4e33-bb18-191ce730720d" for claim "monitoring/elasticsearch-data-monitoring-es-default-0" created
I1221 17:50:49.259912       1 controller.go:644] volume "pvc-4869d0fe-a082-4e33-bb18-191ce730720d" for claim "monitoring/elasticsearch-data-monitoring-es-default-0" saved
I1221 17:50:49.259952       1 controller.go:680] volume "pvc-4869d0fe-a082-4e33-bb18-191ce730720d" provisioned for claim "monitoring/elasticsearch-data-monitoring-es-default-0"
I1221 17:51:20.625165       1 leaderelection.go:202] stopped trying to renew lease to provision for pvc monitoring/elasticsearch-data-monitoring-es-default-0, timeout reached
```

**Comment by balchua** [comment]
@joshbranham is your cluster a single node or multi node?
Thanks

**Comment by joshbranham** [comment]
> @joshbranham is your cluster a single node or multi node?
> Thanks

We are running single node, sorry should have specified :)

**Comment by balchua** [comment]
@joshbranham thanks for the info.
Just to clarify, this happens only when you first create the pvc and apply the manifests?

**Comment by joshbranham** [comment]
> @joshbranham thanks for the info.
> Just to clarify, this happens only when you first create the pvc and apply the manifests?

Yeah it appears to get hung up, it does not happen every time, but we bring up and tear down microk8s single node clusters daily and notice this from time to time. We also notice sometimes it just taking a long time for the PVC to be ready, pods will take upwards of 5min to schedule pending the PVC. Its possible the above logs would have eventually rendered a successful PVC but I re-created them after >5min of waiting

**Comment by skytodmoon** [comment]
microk8s-version: v1.19.15
I met the same problem, If I create 1 master node and 1 worker node, the pvc file will be created under the worker node, so the master node can not find it.

**Comment by neoaggelos** [comment]
This has been fixed for MicroK8s versions 1.21 and newer. The hostpath storage addon will create the volume and set the permissions correctly on the correct cluster node.

Note that while functional in multi-node environments, the hostpath storage addon is not recommended for use in production clusters (for example; migrating pods to a different node will not work because of the volume affinity). For production environments, OpenEBS, MayaStor or any production-grade CSI storage provider should be used instead.
